File: Python/algorithms/sorting/mergeSort.py
"""
Algorithm: Merge Sort

Description:
    Merge Sort is a Divide and Conquer sorting algorithm. It divides the input 
    array into two halves, recursively sorts them, and then merges the sorted halves 
    to produce a fully sorted array.

Steps:
    1. Divide the array into two halves until each subarray has one element.
    2. Recursively sort both halves.
    3. Merge the sorted halves to form a larger sorted array.

Use Case:
    Merge Sort is useful when stable sorting is required or when working with 
    linked lists, external sorting (large data that doesn’t fit into memory), 
    and scenarios where worst-case performance guarantees are important.

Time Complexity:
    - Best Case: O(N log N)
    - Average Case: O(N log N)
    - Worst Case: O(N log N)
    
Space Complexity:
    O(N) auxiliary space (due to temporary arrays during merging).

Stability:
    Merge Sort is a **stable sorting algorithm**, meaning it maintains the 
    relative order of equal elements.
"""
def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2  
        L = arr[:mid]        
        R = arr[mid:]

        # Recursive calls to sort both halves
        merge_sort(L)
        merge_sort(R)

        i = j = k = 0

        # Merge the two halves
        while i < len(L) and j < len(R):
            if L[i] <= R[j]:   
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        # Copy any remaining elements of L
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        # Copy any remaining elements of R
        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1



if __name__ == "__main__":
    arr = [38, 27, 43, 3, 9, 82, 10]
    print("Original array:", arr)
    merge_sort(arr)
    print("Sorted array:", arr)
# Sorted array: [3, 9, 10, 27, 38, 43, 82]

File: JavaScript/algorithms/sorting/mergesort.js
/* Merge Sort

Description: Merge sort is a divide-and-conquer algorithm that divides the input array into two halves, calls itself for the two halves, and then merges the two sorted halves. The merge() function is used for merging two halves.

Time Complexity: O(n log n)

Space Complexity: O(n)
*/

function mergeSort(arr) {
  if (arr.length <= 1) {
    return arr;
  }

  const middle = Math.floor(arr.length / 2);
  const leftList = arr.slice(0, middle);
  const rightList = arr.slice(middle);

  return merge(mergeSort(leftList), mergeSort(rightList));
}

function merge(leftList, rightList) {
  let result = [];
  let left = 0;
  let right = 0;

  while (left < leftList.length && right < rightList.length) {
    if (leftList[left] < rightList[right]) {
      result.push(leftList[left]);
      left++;
    } else {
      result.push(rightList[right]);
      right++;
    }
  }

  return result.concat(leftList.slice(left)).concat(rightList.slice(right));
}

function runTests() {
  console.log("---------- Merge Sort Test Cases ----------");

  // Test 1: Regular unsorted array
  const test1 = [64, 34, 25, 12, 22, 11, 90];
  console.log("\nTest 1 - Original:", test1);
  console.log("Test 1 - Sorted:", mergeSort(test1));

  // Test 2: Already sorted array
  const test2 = [1, 2, 3, 4, 5];
  console.log("\nTest 2 - Original:", test2);
  console.log("Test 2 - Sorted:", mergeSort(test2));

  // Test 3: Reverse sorted array
  const test3 = [5, 4, 3, 2, 1];
  console.log("\nTest 3 - Original:", test3);
  console.log("Test 3 - Sorted:", mergeSort(test3));

  // Test 4: Array with duplicates
  const test4 = [3, 1, 4, 1, 5, 9, 2, 6, 5];
  console.log("\nTest 4 - Original:", test4);
  console.log("Test 4 - Sorted:", mergeSort(test4));

  // Test 5: Single element array
  const test5 = [42];
  console.log("\nTest 5 - Original:", test5);
  console.log("Test 5 - Sorted:", mergeSort(test5));

  // Test 6: Empty array
  const test6 = [];
  console.log("\nTest 6 - Original:", test6);
  console.log("Test 6 - Sorted:", mergeSort(test6));

  // Test 7: Large array
  const test7 = [38, 27, 43, 3, 9, 82, 10];
  console.log("\nTest 7 - Original:", test7);
  console.log("Test 7 - Sorted:", mergeSort(test7));
}

runTests();
